STYLEGUIDE.md

```diff
@@ -64,6 +64,62 @@ Fallback: https://github.com/styleguide/ruby
     first_method.
     second_method
   ```
+  
+### Writing Unit Test
+Our unit tests testing library is [Minitest](https://github.com/minitest/minitest). 
+
+When writing unit test we prefer to use [Specs Syntax](https://github.com/minitest/minitest?tab=readme-ov-file#specs-). Our default MiniTest Spec has been extended to include some RSpec features such as [`subject`](https://rspec.info/features/3-13/rspec-core/subject/explicit-subject/) or [`let!`](https://rspec.info/features/3-13/rspec-core/helper-methods/let/)
+
+#### General Guidelines
+- Group-related tests using `describe` blocks.
+- Use `let` to define memoized helper methods and lazy-loaded variables.
+- Use `before` blocks for setup tasks that need to be run before each example.
+- Use `subject` to explicitly define the value that is returned by the subject method in the example scope.
+- Prefer `.must_` methods over `expect` or `assert_` syntax for assertions, for consistency. See all possible expectations [here](https://github.com/minitest/minitest/blob/v5.18.0/lib/minitest/expectations.rb).
+
+You can find more detailed guidelines and examples of best practices at [betterspecs.org](https://www.betterspecs.org/). 
+
+#### Example Structure
+- Describe the method or class being tested. Use the Ruby documentation convention of `.` (or `::`) when referring to a class method's name and # when referring to an instance method's name.
+- Use context blocks to describe different states or conditions. When describing a context, start its description with 'when', 'with' or 'without'.
+- Write test cases using `it` blocks for individual behaviours.
+
+#### Example
+```ruby
+describe MyClassTest do
+  # `described_class` is a helper method that returns the class or module that is currently being tested, e.g. `MyClass` is returned when described_class is used in `MyClassTest`.
+  let(:described_instance) {described_class.new}
+
+  before do
+    # Setup code
+  end
+
+  describe '#my_method' do
+    subject(:my_method) {described_instance.my_method(argument)}
+
+    let(:argument) {'argument'}
+
+    context 'when condition A is met' do
+      let(:argument) {'argument_a'}
+
+      it 'returns expected result' do
+        # `_my_method` is a helper method assertion method defined for the test `subject`.
+        _my_method.must_equal 'expected result for argument_a'
+      end
+    end
+
+    context 'when condition B is met' do
+      let(:argument) {'argument_b'}
+
+      it 'returns another result' do
+        _my_method.must_equal 'expected result for argument_b'
+      end
+    end
+  end
+end
+```
+
+See a real example in our codebase in the `Queries::SectionTest` [test](https://github.com/code-dot-org/code-dot-org/blob/012a3bcb13f4a265d6570b8995ae872e92a0ab42/dashboard/test/lib/queries/section_test.rb) for the `Queries::Section` [Module](https://github.com/code-dot-org/code-dot-org/blob/012a3bcb13f4a265d6570b8995ae872e92a0ab42/dashboard/lib/queries/section.rb).
```

Comment:
Another example is [omniauth_callbacks_controller_test.rb](https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/test/controllers/omniauth_callbacks_controller_test.rb). It has some more complex examples. It also demonstrates how you can just add in spec-style tests even if there are other tests in the same class.